nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/test/java/org/apache/nifi/processors/azure/storage/ITFetchAzureDataLakeStorage.java

```diff
@@ -16,43 +16,329 @@
+        String fileContent2 = "ContentOfFile2";
+        String inputFlowFileContent = "InputFlowFileContent";
+
+        createDirectoryAndUploadFile(directory, filename1, fileContent1);
+        createDirectoryAndUploadFile(directory, filename2, fileContent2);
+
+        // WHEN
+        // THEN
+        testSuccessfulFetch(this.fileSystemName, directory, filename1, inputFlowFileContent, fileContent1);
+        runner.clearProvenanceEvents();
+        runner.clearTransferState();
+        testSuccessfulFetch(this.fileSystemName, directory, filename2, inputFlowFileContent, fileContent2);
+    }
+
+    @Test
+    public void testFetchFileCaseSensitiveDirectoryName() {
+        // GIVEN
+        String directory1 = "TestDirectory";
+        String directory2 = "Testdirectory";
+        String filename1 = "testFile1.txt";
+        String filename2 = "testFile2.txt";
+        String fileContent1 = "ContentOfFile1";
+        String fileContent2 = "ContentOfFile2";
+        String inputFlowFileContent = "InputFlowFileContent";
+
+        createDirectoryAndUploadFile(directory1, filename1, fileContent1);
+        createDirectoryAndUploadFile(directory2, filename2, fileContent2);
+
+        // WHEN
+        // THEN
+        testSuccessfulFetch(this.fileSystemName, directory1, filename1, inputFlowFileContent, fileContent1);
+        runner.clearProvenanceEvents();
+        runner.clearTransferState();
+        testSuccessfulFetch(this.fileSystemName, directory2, filename2, inputFlowFileContent, fileContent2);
+    }
+
+    @Test
+    public void testFetchFileFromDeepDirectoryStructure() {
+        // GIVEN
+        String directory= "Directory01/Directory02/Directory03/Directory04/Directory05/Directory06/Directory07/"
+                + "Directory08/Directory09/Directory10/Directory11/Directory12/Directory13/Directory14/Directory15/"
+                + "Directory16/Directory17/Directory18/Directory19/Directory20/TestDirectory";
+        String filename = "testFile.txt";
+        String fileContent = "AzureFileContent";
+        String inputFlowFileContent = "InputFlowFileContent";
+
+        createDirectoryAndUploadFile(directory, filename, fileContent);
+
+        // WHEN
+        // THEN
+        testSuccessfulFetch(this.fileSystemName, directory, filename, inputFlowFileContent, fileContent);
+    }
+
+    @Test
+    public void testFetchDirectory() {
+        // GIVEN
+        String parentDirectory = "ParentDirectory";
+        String childDirectory = "ChildDirectory";
+        String filename = "testFile.txt";
+        String fileContent = "AzureFileContent";
+        String inputFlowFileContent = "InputFlowFileContent";
+        String expectedFlowFileContent = "";
+
+        createDirectoryAndUploadFile(parentDirectory + "/" + childDirectory, filename, fileContent);
+
+        // WHEN
+        // THEN
+        testSuccessfulFetch(this.fileSystemName, parentDirectory, childDirectory, inputFlowFileContent, expectedFlowFileContent);
```

Comment:
@MuazmaZ When the specified filename points to a directory (instead of a file) on ALDS, the fetch is successful. The client does not get any error or warning from Azure which would indicate that a directory has been fetched and the file itself does not exist.

Is this the expected behaviour? If yes, is there a way to determine whether a directory or an empty file has been fetched?

Reply:
@turcsanyip The getClient methods don’t perform any network calls so there is no way to know if the path points to a file/directory. If you’d like to confirm if a given resource is a file/directory you could do a getProperties call and check PathProperties.isDirectory.

Reply:
@MuazmaZ Thanks for the info!
I see this `PathProperties.isDirectory` is available from 12.1.x.
I'll create a follow-up jira ticket to clarify how to handle it on our side. Till then this test case should be ignored.